src/http/http.js

```diff
@@ -68,40 +68,45 @@ import {
  *    .catch(err => console.log('An error occurred!', err))
  */
 
-function http (endpoint, { 
-  onSuccess=identity, 
-  onFailure=identity,
-  camelizeResponse=true,
-  before=noop,
-  successDataPath,
-  failureDataPath='errors',
+async function http (endpoint, { 
+  before=noop, 
+  __mock_response, // used for unit testing
   ...options
 }={}) {
-  // Apply config middleware, modifying the options in sequence
-  return composeMiddleware(
+
+  const parsedOptions = await composeMiddleware(
     before,
     setDefaults,
     setAuthHeaders,
     addRootToEndpoint,
     addQueryToEndpoint,
     serializeBody,
     includeCSRFToken,
-    filterFetchOptions,
+    extractFetchOptions,
   )({ endpoint, ...options })
-    // Make the fetch call
-    .then(({ endpoint, ...config }) => fetch(endpoint, config))
+
+  const {
+    onSuccess, 
+    onFailure,
+    camelizeResponse,
+    successDataPath,
+    failureDataPath,
+    endpoint: url,
+    fetchOptions,
+  } = parsedOptions
+
+  try {
```

Comment:
Shouldn't this include the `await composeMiddleware()` step?

Reply:
I don't think it'd work since `onFailure` comes from that step

Reply:
What happens then if `composeMiddleware` fails?

Reply:
Error throws, onFailure isn't called.

Reply:
I guess throwing the identity of the error is the same as letting the thrown error go unhandled?

Reply:
Yep. That wouldn't happen in the `composeMiddleware` case though if that's what you're wondering.